### PR TITLE
[FLINK-3942] [tableAPI] Add support for INTERSECT

### PR DESCRIPTION
**Pull request by wuchong**
Internally, I translate INTERSECT into a Join on all fields and then a distinct for removing duplicate records. 

As Calcite SQL Parser doesn't support `INTERSECT ALL` , so I didn't add `intersectAll()` function to `Table`.

I can add the corresponding documents if needed.

### CODE REVIEW

**Comment by mushketyk** [comment]
I think you need to update documentation for this page: https://ci.apache.org/projects/flink/flink-docs-master/apis/table.html regarding newly added operations.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    val joinFun = new FlatJoinRunner[Any, Any, Any](
+      genFunction.name,
+      genFunction.code,
+      genFunction.returnType)
+
+    val joinOpName = s"intersect: ($intersectSelectionToString)"
+
+    val leftKeys = 0 until left.getRowType.getFieldCount
+    val rightKeys = 0 until left.getRowType.getFieldCount
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: should be `right` instead of `left` but both inputs need to have the same number of fields anyway.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    val intersectDS = leftDataSet.join(rightDataSet).where(leftKeys: _*).equalTo(rightKeys: _*)
+        .`with`(joinFun).name(joinOpName)
+
+    if (all) {
+      intersectDS
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: An `INTERSECT ALL` cannot be executed as a join because a join builds the Cartesian product within a join key. However, `INTERSECT ALL` should return as many identical records as are in both inputs, i.e., with `A = [1, 1, 1, 2, ,2]` and `B = [1, 2, 2, 3]` `A INTERSECT ALL B` should return `[1, 2, 2]`. Using a join we'd get `[1, 1, 1, 2, 2, 2, 2]`.

`INTERSECT ALL` can be executed with a `CoGroup` which alternately forwards both iterators and only emits a record if both iterators had an element. We do not need a code-generated function for this, as the records are not modified but simply forwarded.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  override def translateToPlan(
+      tableEnv: BatchTableEnvironment,
+      expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
+
+    var leftDataSet: DataSet[Any] = null
+    var rightDataSet: DataSet[Any] = null
+
+    expectedType match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: We can do the type conversion after the co-group with a map.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    val leftKeys = 0 until left.getRowType.getFieldCount
+    val rightKeys = 0 until left.getRowType.getFieldCount
+
+    val intersectDS = leftDataSet.join(rightDataSet).where(leftKeys: _*).equalTo(rightKeys: _*)
+        .`with`(joinFun).name(joinOpName)
+
+    if (all) {
+      intersectDS
+    } else {
+      intersectDS.distinct()
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: I would also use a coGroup for `INTERSECT` (without `ALL`) because we would not need a distinct afterwards. It's the more robust choice because it won't create a huge intermediate result in case of many duplicate records.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 5]
```diff
@@ -446,6 +446,31 @@ class Table(
   }
 
   /**
+    * Intersect two [[Table]]s with duplicate records removed. Intersect returns rows only
+    * from the left table that are identical to a row in the right table.
+    * Similar to an SQL INTERSECT. The fields of the two union operations must fully overlap.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: "The fields of the two union operations must fully overlap." -> "The fields of the two intersected tables must fully overlap."

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 6]
```diff
@@ -446,6 +446,31 @@ class Table(
   }
 
   /**
+    * Intersect two [[Table]]s with duplicate records removed. Intersect returns rows only
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: The description of intersect suggests that the same row is emitted twice if it appears two times in the left table which is not correct. Rows that appear in both inputs are emitted exactly once.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/table/IntersectITCase.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.flink.api.table.Table;
+import org.apache.flink.api.table.TableEnvironment;
+import org.apache.flink.test.javaApiOperators.util.CollectionDataSets;
+import org.apache.flink.test.util.MultipleProgramsTestBase;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class IntersectITCase extends MultipleProgramsTestBase {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/table/IntersectITCase.java]: We try to keep the number of integration tests for the Table API to a minimum to not exceed Travis' build time threshold of 2 hours. Actually, we try to remove as many as possible without cutting back test coverage. The Java Table API tests for intersect overlap with the Scala Table API and can be removed, IMO.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/IntersectITCase.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import scala.util.Random
+
+@RunWith(classOf[Parameterized])
+class IntersectITCase(
+    mode: TestExecutionMode,
+    configMode: TableConfigMode)
+  extends TableProgramsTestBase(mode, configMode) {
+
+  @Test
+  def testIntersect(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/IntersectITCase.scala]: I think we need only this test to verify that intersect works correctly. However, for each test class (child of `TableProgramsTestBase`) a new test cluster is started which causes significant overhead. Can you add this test to `UnionITCase` and rename it to `SetOperatorsITCase`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala [hunk 9]
```diff
@@ -260,10 +260,14 @@ class CodeGenerator(
     val input1AccessExprs = for (i <- 0 until input1.getArity)
       yield generateInputAccess(input1, input1Term, i)
 
-    val input2AccessExprs = input2 match {
-      case Some(ti) => for (i <- 0 until ti.getArity)
-        yield generateInputAccess(ti, input2Term, i)
-      case None => Seq() // add nothing
+    val input2AccessExprs = if (input1.getArity < resultFieldNames.length) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala]: Can this change be reverted if we use cogroup and an optional converter map? Or would you consider this as a bug fix?

**Comment by fhueske** [comment]
Hi @wuchong, thanks for the PR. I think it is better to use a CoGroup instead of a Join for `INTERSECT` (even though, I said "Join" in the original JIRA). For `INTERSECT ALL`, CoGroup is required. It would also be good to update the Table API and SQL documentation as @mushketyk suggested. 
Thanks, Fabian

**Comment by wuchong** [comment]
Hi @fhueske,  that's a good idea to use CoGroup instead of Join.  I updated the PR according to your advice. Meanwhile, I updated the document too (correct me if I describe wrong). 

NOTE:

1. use CoGroup instead of Join,  no code gen.
1. add INTERSECT related tests into UnionITCase, and rename it to `SetOperatorsITCase`
2. remove INTERSECT Java API tests 
3. add the `intersectAll` function to `Table`
4. mark `testIntersectAll` as `@ignore` in `sql/SetOperatorsITCase`, because calcite sql parser doesn't support INTERSECT ALL, it will throw the following exception:

```
java.lang.AssertionError: Internal error: set operator INTERSECT ALL not suported

  at org.apache.calcite.util.Util.newInternal(Util.java:777)
  at org.apache.calcite.sql2rel.SqlToRelConverter.convertSetOp(SqlToRelConverter.java:2920)
  at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:2885)
  at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:582)
  at org.apache.flink.api.table.FlinkPlannerImpl.rel(FlinkPlannerImpl.scala:114)
  at org.apache.flink.api.table.BatchTableEnvironment.sql(BatchTableEnvironment.scala:132)
  at org.apache.flink.api.scala.batch.sql.SetOperatorsITCase.testIntersectAll(SetOperatorsITCase.scala:169)
```

File: docs/apis/table.md [hunk 10]
```diff
@@ -535,6 +535,30 @@ Table result = left.unionAll(right);
       </td>
     </tr>
 
+	<tr>
+      <td><strong>Intersect</strong></td>
+      <td>
+        <p>Similar to a SQL INTERSECT clause. Intersects two tables with duplicate records removed. Both tables must have identical field types.</p>
+{% highlight scala %}
```
Comment (fhueske) [inline docs/apis/table.md]: This should be Java code + `highlight java`

File: docs/apis/table.md [hunk 11]
```diff
@@ -535,6 +535,30 @@ Table result = left.unionAll(right);
...
+{% endhighlight %}
+      </td>
+    </tr>
+
+	<tr>
+      <td><strong>IntersectAll</strong></td>
+      <td>
+        <p>Similar to a SQL INTERSECT ALL clause. Intersects two tables. It returns as many identical records as are in both tables. Both tables must have identical field types.</p>
+{% highlight scala %}
+val left = ds1.toTable(tableEnv, "a, b, c");
```
Comment (fhueske) [inline docs/apis/table.md]: This should be Java code + `highlight java`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 12]
```diff
@@ -83,66 +82,68 @@ class DataSetIntersect(
...
+    val coGroupedDs = leftDataSet.coGroup(rightDataSet)
 
+    val leftType = leftDataSet.getType
+    val rightType = rightDataSet.getType
 
-    val body = s"""
-            |${conversion.code}
-            |${generator.collectorTerm}.collect(${conversion.resultTerm});
-            |""".stripMargin
+    // If it is atomic type, the field expression need to be "*".
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: I think you can use `*` for all cases. Composite types should be (recursively) expanded to all atomic member types.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala [hunk 13]
```diff
@@ -83,66 +82,68 @@ class DataSetIntersect(
...
+    } else {
+      coGroupedDs.where("*")
+    }
 
-    val genFunction = generator.generateFunction(
-      ruleDescription,
-      classOf[FlatJoinFunction[Any, Any, Any]],
-      body,
-      returnType)
+    val coGroupedWithoutFunctionDs =
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala]: Same here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 14]
```diff
@@ -446,9 +446,9 @@ class Table(
   }
 
   /**
-    * Intersect two [[Table]]s with duplicate records removed. Intersect returns rows only
-    * from the left table that are identical to a row in the right table.
-    * Similar to an SQL INTERSECT. The fields of the two union operations must fully overlap.
+    * Intersect two [[Table]]s with duplicate records removed. Intersect returns records that
+    * exist in both tables, and emit exactly once. Similar to an SQL INTERSECT. The fields of
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: I would rephrase this to "Intersect returns records that exist in both tables. If a record is present one or both tables more than once, it is returned just once, i.e., the resulting table has no duplicate records."

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 15]
```diff
@@ -471,6 +471,32 @@ class Table(
   }
 
   /**
+    * Intersect two [[Table]]s. Intersect All returns records that exist in both tables, but not
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: I would rephrase this to "Intersect All returns records that exist in both tables. If a record is present in both tables more than once, it is returned as many times as it is present in both tables, i.e., the resulting table might have duplicate records."

**Comment by fhueske** [comment]
Hi @wuchong, thanks for the update! The PR looks mostly good. I left a few minor comments.

Thanks, Fabian

**Comment by wuchong** [comment]
Thanks @fhueske for your review,  I have addressed all the comments and squashed the commit.

**Comment by twalthr** [comment]
@wuchong Fabian is on vacation at the moment. I will review your code the next days.

**Comment by wuchong** [comment]
@twalthr   Thanks a lot.

**Comment by twalthr** [comment]
Thanks for the contribution @wuchong! I have fixed some typos but the PR looks good to merge.

Merging...
